### PR TITLE
Update API guidelines with additional rules including SemVer

### PR DESCRIPTION
**Pull request by dthaler**
Addresses issue #2543
Also addresses remaining suggestions on PR #2163 as discussed by the
CGC.

Signed-off-by: Dave Thaler <dthaler@microsoft.com>

### CODE REVIEW

File: docs/ApiGuidelines.md [hunk 1]
```diff
@@ -46,8 +55,10 @@ Experimental API requirements:
 * Experimental APIs must be in a header file that is published to the SDK output directory. They
   can be in their own header, or in a header file that is shared with public APIs but only
   if surrounded by an appropriate ifdef (e.g., "#ifdef OE\_CONTEXT\_SWITCHLESS\_EXPERIMENTAL\_FEATURE").
-* Experimental APIs should be fully documented using doxygen markup in the header file
-* Experimental APIs must have at least one test case.
+* Experimental APIs should be fully documented using doxygen markup in the header file, and include the @experimental markup.
```
Comment (CodeMonkeyLeet) [inline docs/ApiGuidelines.md]: Per the discussion in PR [#2579](https://github.com/openenclave/openenclave/pull/2579#discussion_r384115879), do we want to be more prescriptive in the use of the `@experimental` API tag? For example, do we want to recommend that all new APIs should be marked experimental APIs for at least a minor version release?

File: docs/ApiGuidelines.md [hunk 2]
```diff
@@ -46,8 +55,10 @@ Experimental API requirements:
 * Experimental APIs must be in a header file that is published to the SDK output directory. They
   can be in their own header, or in a header file that is shared with public APIs but only
   if surrounded by an appropriate ifdef (e.g., "#ifdef OE\_CONTEXT\_SWITCHLESS\_EXPERIMENTAL\_FEATURE").
```
Comment (CodeMonkeyLeet) [inline docs/ApiGuidelines.md]: Nit: consider using code markdown for readability and not having to use character escapes (e.g. `#ifdef OE_CONTEXT_SWITCHLESS_EXPERIMENTAL_FEATURE`)

File: docs/ApiGuidelines.md (hunk 1 above)
Comment (radhikaj) [reply docs/ApiGuidelines.md]: I think this is a good idea. Can we add it to the follow up PR that will address the outstanding comments here?

File: docs/ApiGuidelines.md [hunk 3]
```diff
@@ -14,16 +14,25 @@ Public APIs are the supported APIs that apps can use.
 
 Public API requirements: 
 
-* All public APIs must have at least one test case.
```
Comment (radhikaj) [inline docs/ApiGuidelines.md]: I would like to add that any new public API should have a design PR that precedes it. Once it is approved, contributors may submit code to implement the new public API.

File: docs/ApiGuidelines.md (hunk 1 above)
Comment (achamayou) [reply docs/ApiGuidelines.md]: I don't think that serves any purpose pre-1.0. There's no effective compatibility SLA, and staying away from `@experimental` APIs does nothing to change that. If OE decides to break compatibility, that's a minor release, and there's no break-free upgrade path from the previous release at all.

**Comment by jhand2** [comment]
Merging this. Other discussions can continue in issues

bors r+

**Pull request merged**